docs/dynamic-component-loading.md

```diff
@@ -0,0 +1,91 @@
+* MSBuild tasks - all tasks in MSBuild are dynamically loaded. Some tasks come with additional dependencies which can collide with each other or MSBuild itself as well.
+* Roslyn analyzers - similar to MSBuild tasks, the Roslyn compiler dynamically loads analyzers which are separate components with potentially conflicting dependencies.
+* XUnit loading tests - the test runner acts as an app and the test is loaded dynamically. The test can have any number of dependencies. Finding and resolving those dependencies is challenging.
+* ASP .NET's `dotnet watch` - ability to dynamically reload an app without restarting the process. Each version of the app is inherently in collision with any previous version. The old version should be unloaded.
+
+In lot of these cases the component which is to be loaded dynamically has a non-trivial set of dependencies which are unknown to the app itself. So the loading mechanism has to be able to resolve them.
+
+## Declaring dependencies
+The .NET Core SDK tooling produces `.deps.json` which is a dependency manifest for dotnet apps. It enables components (or apps) to load dependencies from locations other than the base directory (ex: from packages, platform-specific publish directories, etc.)  
+.NET Core first builds a set of directories to look for binaries (called ProbingPaths) based on application/component base directory, framework directory, `.runtimeconfig.dev.json`, command line, servicing locations, etc. For more details see [host-probing](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/host-probing.md).
+
+For an application or component, the `.deps.json` file specifies:
+* A set of binary dependencies -- which are searched within ProbingPaths
+* Relative paths to locate them -- relative paths in `.deps.json` file can be used to locate architecture-specific dependencies within publish/package directories.
+
+If the app depends on any frameworks, the `.deps.json` files of those framework are similarly processed.
+Further details about the algorithm used for processing dependencies can be found in [assembly-conflict-resolution](
+https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/assembly-conflict-resolution.md).
+
+## Dynamic loading with dependencies
+We propose to add a new public API which would dynamically load a component with these properties:
+* Component is loaded in isolation from the app (and other components) so that potential collisions are not an issue
+* Component can use `.deps.json` to describe its dependencies. This includes the ability to describe additional NuGet packages, RID-specific and/or native dependencies
+* Component can chose to rely on the app for certain dependencies by not including them in its `.deps.json`
+* Optionally such component can be enabled for unloading
+
+Public API (early thinking):
+```csharp
+class Assembly
+{
+    public static Assembly LoadFileWithDependencies(string path);
```

Comment:
@jeffschwMSFT asked about loading assemblies (and their dependencies) to an existing load-context, typically the default load context. If this is a typical use case for developers, we should support it, via an isolation switch.

```csharp
class Assembly
{
    public static Assembly LoadFileWithDependencies(string path, bool Isolated);
}
```
I don't know of any technical reasons not to support this behavior.

Reply:
How would this API handle conflicts? Fail right away if there are any conflicts?

Reply:
80% of existing use of ALCs load within the default.  I think we should consider this as an append to the end scenario.  If there is a conflict and the existing TPA cannot resolve it, then it would fail to load.  We could either do that proactively or fail at the point of the load of the dependent assembly.

Reply:
I'm just curious: How did we learn that 80% of ALC use cases are loading into default?
Also - what do you mean by that: That in the Load the ALC's code calls something like Assembly.LoadFrom?
Also - do we know if that's the intended behavior people wanted?

Do you have a sample I can look at - I would be really interested why would somebody do this...

Reply:
> 80% of existing use of ALCs load within the default.

Do you have a real world example that would be a good candidate for using to `LoadFileWithDependencies` into default context?

Reply:
> 80% of existing use of ALCs load within the default

Does this really mean "80% of existing use of ALCs load dependencies within the default ALC"

This would also make them fail to unload....

Reply:
If you do a search for ALC in GitHub or nuget you will find that AssemblyLoadContext.Default is the primary use case.  In addition extrapolating from how AppDomains are used (or more often not) leads me to believe that the non-isolated case is likely the primary case.

Reply:
For me in reading the existing issues in this space, the #1 issue is people are having problems doing dynamic loads.  Unload ability and isolation are secondary issues.

> Do you have a real world example that would be a good candidate for using LoadFileWithDependencies into default context API?

Most any dynamic load scenario.

Reply:
Looking through the various samples and use cases of default ALC - to me it seems the default ALC usage is used a fallback. In this case explicit fallback (where returning null would do the job as well).
I absolutely agree with the fact that # 1 problem is dynamic loads are too hard. But I would argue whether people explicitly want to load into Default.
They probably just want it to work. Using the default is likely a result of debugging, where it failed to load some basic dependency (something from a framework or so), and so the code redirects to default which already has that.

The problem I see with allowing a .deps.json based resolver to load directly into Default is that it is VERY likely to fail. Basically once the plugin has any assembly with the same name as the app, it will fail (even the same exact version, this is because of the current binder behavior which we can't really change). This doesn't feel like a "pit of success" - it seems to be quite the opposite.

Back to the "dynamic load is hard": I think there are 2 main reasons why it's hard:
* If there are dependencies, there's no built-in way to resolve them (other then Assembly.LoadFrom) - this has little to do with loading into default, it's just a lack of functionality.
* Avoiding the first problem can be done by directly loading into Default which has all the dependencies, but now comes the compatibility problem, where it's really easy to fail to load due to version (or other) mismatches. - this is about isolation. We can't solve the collisions without isolation. If we would, we would go back to desktop which allowed resolving higher version reference to a lower version instance - which is really bad.

I haven't yet seen a scenario where loading directly into default with all of the dependency tree is desirable. But there definitely might be one. I'm not against adding an option to do so, but I think it should not be the default. By default we should load with isolation on.

Reply:
> search for ALC in GitHub or nuget you will find that AssemblyLoadContext.Default

I have looked through first several pages of https://github.com/search?l=C%23&q=AssemblyLoadContext.Default&type=Code . I have not found a case that would be a good candidate for loading with isolation off.

Reply:
Leaving the default decision aside - I think that customer feedback can help us resolve that.  I do believe there are a lot of cases where customers would rather just have all their code in the same context.  It makes code sharing doable.  As it stands today, the only way to communicate between contexts is via reflection with only runtime types that exchange.  I have seen very few examples of people getting that right.  Once we have a reasonable higher level communication strategy designed, I could see a path ahead that enabled our customers to get isolation (which I agree is desirable) and allow for sharing in a reasonable way.

Reply:
Loading binaries to the parent context is our basic unit of sharing. So, some apps may want to load dependencies (or dependent frameworks) into their own context instead of creating an ALC just to act as a sharing layer [Is this operation significant enough to impact performance?]

Similar behavior can be achieved using existing mechanisms, via a cooperative protocol with the plugin. But LoadWithDependencies() call is more straightforward.

I don't have a strong preference for the isolation switch, we'll only know based on use cases.

Reply:
> How would this API handle conflicts? Fail right away if there are any conflicts? 

Yes, the behavior should be identical to loading different versions of the same DLL into an ALC, so I think the load operation should fail right away.

Reply:
We should probably call the switch something other than "isolation" because loads in the plugin ALC can cause things to be loaded into the parent ALC, which can affect other plugin ALCs.

Reply:
>  I think that customer feedback can help us resolve that

Agree. It would be very useful to get the first iteration on these APIs (the set that we all agree about) into the first .NET Core 3.0 preview to start getting feedback as soon as possible.

Reply:
> As it stands today, the only way to communicate between contexts is via reflection with only runtime types that exchange. 

This is not the case. The component can choose to share assembly with the app, by not including it in its deployment. There's tooling for this on VS side already (CopyLocal setting for assembly references, similar exists for project and Nuget references). CLI doesn't have this as an option … yet. We might want to add it if we think there's a need. Otherwise modifying the proj file also works.

For example tasks for MSBuild. The plugin (task) implements the `ITask` interface which comes from `Microsoft.Build.Framework.dll`. It makes perfect sense to not include the `Microsoft.Build.Framework.dll` with the plugin, as that's clearly part of the host (MSBuild). Doing that will also inherently share all the types from that assembly between the host and all the plugins.

Reply:
WRT customer feedback.  I hope there are multiple avenues to get early feedback: let's move this discussion to coreclr and invite people with existing feedback to chime in, talk with Richard to get early enterprise'ish usage, and get them into preview for feedback.

Reply:
>>As it stands today, the only way to communicate between contexts is via reflection with only runtime types that exchange.

>This is not the case. The component can choose to share assembly with the app, by not including it in its deployment. 

Perhaps we are having a difference in words only... but for me CreateInstance is not a natural pattern and fits within my reflection use case.  If we had a LoadAndUnwrap then perhaps that is pretty natural, as that does not require any prior knowledge of what is loaded where and you get an 'object' back.

Reply:
My 2 cents here is that isolation should be the default since that seems to be the entire point of this feature - but perhaps not on this API. I am generally leery of APIs that take `bool` arguments because any API that takes a `bool` is invariably trying to make a decision about what the default behavior. In this case adding a new function that indicates `Assembly.LoadIsolated()` makes way more sense to me and is clear about the very new semantics we are providing.

Reply:
+1 on this API in essence being the 'append to an existing ALC' by loading with dependencies and the ALC API being the way to load with isolation.  The more general scenario that I hope we are able to enable is loading multiple plug-ins into the same ALC (be it either default or a new one).  Having a 1:1 relationship of a deps to a ALC (which is what isolation by default in essence turns into) I feel is too limiting.

Reply:
I am wondering whether it would be better to drop the simple convenience API from the first phase given that we do not have clear answer for the right default. We can always introduce it later once we have better data about what the default should be and we have somebody asking for it.